### PR TITLE
ASoC: rt715-sdca: First version of rt715 sdw sdca codec driver.

### PR DESCRIPTION
**Pull request by jack-cy-yu**
First version of rt715 sdw sdca codec driver.

### CODE REVIEW

File: sound/soc/codecs/rt715-sdca.c [hunk 1]
```diff
@@ -0,0 +1,943 @@
...
+static int rt715_index_read(struct rt715_priv *rt715,
+		unsigned int nid, unsigned int reg, unsigned int *value)
+{
+	struct regmap *regmap = rt715->regmap;
+	unsigned int addr;
+	int ret;
+
+	addr = ((nid << 20) | reg);
+
+	*value = 0;
```
Comment (RanderWang) [inline sound/soc/codecs/rt715-sdca.c]: why it is needed ? if ret == 0,  the value can be incorrect ?

File: sound/soc/codecs/rt715-sdca.c [hunk 2]
```diff
@@ -0,0 +1,943 @@
...
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct soc_mixer_control *mc =
+		(struct soc_mixer_control *)kcontrol->private_value;
+	struct rt715_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int val_l, val_r, gain_l_val, gain_r_val;
+	unsigned int i;
+
+	/* control value to 2s complement */
+	/* L channel */
```
Comment (RanderWang) [inline sound/soc/codecs/rt715-sdca.c]: refine comments style

File: sound/soc/codecs/rt715-sdca.c [hunk 3]
```diff
@@ -0,0 +1,943 @@
...
+#include "rt715-sdw-sdca.h"
+
+static int rt715_index_write(struct rt715_priv *rt715, unsigned int nid,
+		unsigned int reg, unsigned int value)
+{
+	struct regmap *regmap = rt715->regmap;
+	unsigned int addr;
+	int ret;
+
+	addr = ((nid << 20) | reg);
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: superfluous external parentheses

File: sound/soc/codecs/rt715-sdca.c [hunk 4]
```diff
@@ -0,0 +1,943 @@
...
+	gain_l_val = ucontrol->value.integer.value[0];
+	if (gain_l_val > mc->max)
+		gain_l_val = mc->max;
+	val_l = gain_l_val;
+
+	if (mc->shift == 8)
+		gain_l_val = (gain_l_val * 10) << mc->shift;
+	else {
+		if (gain_l_val > mc->shift)
+			gain_l_val = (gain_l_val - mc->shift) * 375;
```
Comment (RanderWang) [inline sound/soc/codecs/rt715-sdca.c]: 375 ? macro define ?

File: sound/soc/codecs/rt715-sdca.c [hunk 5]
```diff
@@ -0,0 +1,943 @@
...
+	unsigned int addr;
+	int ret;
+
+	addr = ((nid << 20) | reg);
+
+	ret = regmap_write(regmap, addr, value);
+	if (ret < 0) {
+		pr_err("Failed to set private value: %08x <= %04x %d\n", ret,
+			addr, value);
+	}
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: superfluous braces

File: sound/soc/codecs/rt715-sdca.c [hunk 6]
```diff
@@ -0,0 +1,943 @@
...
+}
+
+static int rt715_index_read(struct rt715_priv *rt715,
+		unsigned int nid, unsigned int reg, unsigned int *value)
+{
+	struct regmap *regmap = rt715->regmap;
+	unsigned int addr;
+	int ret;
+
+	addr = ((nid << 20) | reg);
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: as above

File: sound/soc/codecs/rt715-sdca.c [hunk 7]
```diff
@@ -0,0 +1,943 @@
...
+	if (mc->shift == 8) {
+		ctl_l = val_l >> mc->shift;
+		ctl_l /= 10;
+	} else {
+		ctl_l = val_l;
+		if (ctl_l & BIT(15)) {
+			ctl_l = val_l - 1;
+			ctl_l = ~ctl_l;
+			ctl_l &= 0xffff;
+			neg_flag = 1;
```
Comment (RanderWang) [inline sound/soc/codecs/rt715-sdca.c]: how about ctl_l = ~(val_l - 1) & 0xffff;

File: sound/soc/codecs/rt715-sdca.c [hunk 8]
```diff
@@ -0,0 +1,943 @@
...
+	if (mc->shift == 8) {
+		ctl_r = val_r >> mc->shift;
+		ctl_r /= 10;
+	} else {
+		ctl_r = val_r;
+		if (ctl_r & BIT(15)) {
+			ctl_r = val_r - 1;
+			ctl_r = ~ctl_r;
+			ctl_r &= 0xffff;
+			neg_flag = 1;
```
Comment (RanderWang) [inline sound/soc/codecs/rt715-sdca.c]: also here ctl_r = ~(val_r -1 ) & 0xffff

File: sound/soc/codecs/rt715-sdca.c [hunk 9]
```diff
@@ -0,0 +1,943 @@
...
+	} else if (strstr(ucontrol->id.name, "ADC 25 Mux")) {
+		mask_sft = 0;
+	} else
+		return -EINVAL;
+
+	/* nid = e->reg, vid = 0xf01 */
+	rt715_index_read(rt715, RT715_VENDOR_HDA_CTL,
+		RT715_HDA_LEGACY_MUX_CTL1, &val);
+	val = (val >> mask_sft) & 0xf;
+
```
Comment (RanderWang) [inline sound/soc/codecs/rt715-sdca.c]: redundant blank line

File: sound/soc/codecs/rt715-sdca.c [hunk 10]
```diff
@@ -0,0 +1,943 @@
...
+{
+	unsigned int tmp, orig;
+	int ret;
+
+	ret = rt715_index_read(rt715, nid, reg, &orig);
+	if (ret < 0)
+		return ret;
+
+	tmp = orig & ~mask;
+	tmp |= val & mask;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: you don't seem to need `orig` any more, so you don't need two variables `orig` and `tmp` here.  Also, you could actually use `set_mask_bits()` here, but that would probably be a bit of an overkill for a local variable

File: sound/soc/codecs/rt715-sdca.c [hunk 11]
```diff
@@ -0,0 +1,943 @@
...
+		gain_l_val = mc->max;
+	val_l = gain_l_val;
+
+	if (mc->shift == 8)
+		gain_l_val = (gain_l_val * 10) << mc->shift;
+	else {
+		if (gain_l_val > mc->shift)
+			gain_l_val = (gain_l_val - mc->shift) * 375;
+		else
+			gain_l_val = (mc->shift - gain_l_val) * 375;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]:
```
gain_l_val = abs(gain_l_val - mc->shift) * 375;
```

File: sound/soc/codecs/rt715-sdca.c [hunk 12]
```diff
@@ -0,0 +1,943 @@
...
+	stream_config.direction = direction;
+
+	num_channels = params_channels(params);
+	port_config.ch_mask = (1 << (num_channels)) - 1;
+	port_config.num = port;
+
+	retval = sdw_stream_add_slave(rt715->slave, &stream_config,
+					&port_config, 1, stream->sdw_stream);
+	if (retval) {
+		dev_err(dai->dev, "Unable to configure port\n");
```
Comment (RanderWang) [inline sound/soc/codecs/rt715-sdca.c]: print err to help debugging

File: sound/soc/codecs/rt715-sdca.c [hunk 13]
```diff
@@ -0,0 +1,943 @@
...
+		gain_l_val = (gain_l_val * 10) << mc->shift;
+	else {
+		if (gain_l_val > mc->shift)
+			gain_l_val = (gain_l_val - mc->shift) * 375;
+		else
+			gain_l_val = (mc->shift - gain_l_val) * 375;
+
+		gain_l_val <<= 8;
+		gain_l_val /= 1000;
+		if (!(val_l > mc->shift)) {
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: `if (val_l <= mc->shift)`

File: sound/soc/codecs/rt715-sdca.c [hunk 14]
```diff
@@ -0,0 +1,943 @@
...
+		return -ENOMEM;
+
+	dev_set_drvdata(dev, rt715);
+	rt715->slave = slave;
+	rt715->regmap = regmap;
+	rt715->sdw_regmap = sdw_regmap;
+	rt715->hw_sdw_ver = slave->id.sdw_version;
+	/*
+	 * Mark hw_init to false
+	 * HW init will be performed when device reports present
```
Comment (RanderWang) [inline sound/soc/codecs/rt715-sdca.c]: refine comments style

File: sound/soc/codecs/rt715-sdca.c [hunk 15]
```diff
@@ -0,0 +1,943 @@
...
+		else
+			gain_r_val = (mc->shift - gain_r_val) * 375;
+
+		gain_r_val <<= 8;
+		gain_r_val /= 1000;
+		if (!(val_r > mc->shift)) {
+			gain_r_val = ~gain_r_val;
+			gain_r_val += 1;
+		}
+		gain_r_val &= 0xffff;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: same comments as for `gain_l_val`

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 16]
```diff
@@ -0,0 +1,479 @@
...
+			return ret;
+		ret = regmap_read(rt715->sdw_regmap,
+			reg | BIT(13), &sdw_data_1);
+		if (ret < 0)
+			return ret;
+		*val = ((sdw_data_1 & 0xff) << 8) | (sdw_data_0 & 0xff);
+		goto _done_;
+	default:
+		break;
+	}
```
Comment (RanderWang) [inline sound/soc/codecs/rt715-sdw-sdca.c]: append a blank line ?

File: sound/soc/codecs/rt715-sdca.c [hunk 17]
```diff
@@ -0,0 +1,943 @@
...
+			regmap_write(rt715->regmap,
+				mc->rreg, gain_l_val);
+		} else {
+			/* Lch*/
+			regmap_write(rt715->regmap,
+				mc->reg, gain_l_val);
+			/* Rch */
+			regmap_write(rt715->regmap,
+				mc->rreg, gain_r_val);
+		}
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: mmh, I don't understand why you need that `if` - just use the generic form as if they weren't equal.

File: sound/soc/codecs/rt715-sdca.c [hunk 18]
```diff
@@ -0,0 +1,943 @@
...
+
+		regmap_read(rt715->regmap, mc->reg, &val_l);
+		regmap_read(rt715->regmap, mc->rreg, &val_r);
+		if (val_r == gain_r_val && val_l == gain_l_val)
+			break;
+	}
+
+
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: superfluous empty line

File: sound/soc/codecs/rt715-sdca.c [hunk 19]
```diff
@@ -0,0 +1,943 @@
...
+	return 0;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: even if all 3 attempts to set gain failed?

File: sound/soc/codecs/rt715-sdca.c [hunk 20]
```diff
@@ -0,0 +1,943 @@
...
+	struct rt715_priv *rt715 = snd_soc_component_get_drvdata(component);
+	unsigned int val_l, val_r, ctl_l, ctl_r, neg_flag = 0;
+
+	regmap_read(rt715->regmap, mc->reg, &val_l);
+	regmap_read(rt715->regmap, mc->rreg, &val_r);
+
+	/* L channel */
+	if (mc->shift == 8) {
+		ctl_l = val_l >> mc->shift;
+		ctl_l /= 10;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: `ctl_l = (val_l >> mc->shift) / 10;`

File: sound/soc/codecs/rt715-sdca.c [hunk 21]
```diff
@@ -0,0 +1,943 @@
...
+	} else {
+		ctl_l = val_l;
+		if (ctl_l & BIT(15)) {
+			ctl_l = val_l - 1;
+			ctl_l = ~ctl_l;
+			ctl_l &= 0xffff;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: `ctl_l = 0xffff & ~(val_l - 1);`

File: sound/soc/codecs/rt715-sdca.c [hunk 22]
```diff
@@ -0,0 +1,943 @@
...
+			neg_flag = 1;
+		}
+		ctl_l *= 1000;
+		ctl_l >>= 8;
+		if (neg_flag)
+			ctl_l = mc->shift - (ctl_l / 375);
+		else
+			ctl_l = mc->shift + (ctl_l / 375);
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: superfluous parentheses twice above

File: sound/soc/codecs/rt715-sdca.c [hunk 23]
```diff
@@ -0,0 +1,943 @@
...
+			ctl_r = ~ctl_r;
+			ctl_r &= 0xffff;
+			neg_flag = 1;
+		}
+		ctl_r *= 1000;
+		ctl_r >>= 8;
+		if (neg_flag)
+			ctl_r = mc->shift - (ctl_r / 375);
+		else
+			ctl_r = mc->shift + (ctl_r / 375);
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: same here as for L channel

File: sound/soc/codecs/rt715-sdca.c [hunk 24]
```diff
@@ -0,0 +1,943 @@
...
+	if (strstr(ucontrol->id.name, "ADC 22 Mux")) {
+		mask_sft = 12;
+	} else if (strstr(ucontrol->id.name, "ADC 23 Mux")) {
+		mask_sft = 8;
+	} else if (strstr(ucontrol->id.name, "ADC 24 Mux")) {
+		mask_sft = 4;
+	} else if (strstr(ucontrol->id.name, "ADC 25 Mux")) {
+		mask_sft = 0;
+	} else
+		return -EINVAL;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: superfluous braces in all `if` branches

File: sound/soc/codecs/rt715-sdca.c [hunk 25]
```diff
@@ -0,0 +1,943 @@
...
+
+	/* nid = e->reg, vid = 0xf01 */
+	rt715_index_read(rt715, RT715_VENDOR_HDA_CTL,
+		RT715_HDA_LEGACY_MUX_CTL1, &val);
+	val = (val >> mask_sft) & 0xf;
+
+
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: 1 empty line, please

File: sound/soc/codecs/rt715-sdca.c [hunk 26]
```diff
@@ -0,0 +1,943 @@
...
+	/*
+	 * The first two indices of ADC Mux 24/25 are routed to the same
+	 * hardware source. ie, ADC Mux 24 0/1 will both connect to MIC2.
+	 * To have a unique set of inputs, we skip the index1 of the muxes.
+	 */
+	if ((e->reg == RT715_MUX_IN3 || e->reg == RT715_MUX_IN4) && (val > 0))
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: no need for parentheses around `val > 0`

File: sound/soc/codecs/rt715-sdca.c [hunk 27]
```diff
@@ -0,0 +1,943 @@
...
+	if (strstr(ucontrol->id.name, "ADC 22 Mux")) {
+		mask_sft = 12;
+	} else if (strstr(ucontrol->id.name, "ADC 23 Mux")) {
+		mask_sft = 8;
+	} else if (strstr(ucontrol->id.name, "ADC 24 Mux")) {
+		mask_sft = 4;
+	} else if (strstr(ucontrol->id.name, "ADC 25 Mux")) {
+		mask_sft = 0;
+	} else
+		return -EINVAL;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: same with braces

File: sound/soc/codecs/rt715-sdca.c [hunk 28]
```diff
@@ -0,0 +1,943 @@
...
+
+	/* Verb ID = 0x701h, nid = e->reg */
+	val = snd_soc_enum_item_to_val(e, item[0]) << e->shift_l;
+
+	rt715_index_read(rt715, RT715_VENDOR_HDA_CTL,
+		RT715_HDA_LEGACY_MUX_CTL1, &val2);
+	val2 = (val2 >> mask_sft) & 0xf;
+
+
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: 1 empty line

File: sound/soc/codecs/rt715-sdca.c [hunk 29]
```diff
@@ -0,0 +1,943 @@
...
+		snd_soc_component_get_dapm(component);
+
+	switch (level) {
+	case SND_SOC_BIAS_PREPARE:
+		break;
+	case SND_SOC_BIAS_STANDBY:
+		break;
+	default:
+		break;
+	}
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: hm, that above `switch` doesn't seem to be doing much :-)

File: sound/soc/codecs/rt715-sdca.c [hunk 30]
```diff
@@ -0,0 +1,943 @@
...
+				int direction)
+{
+
+	struct sdw_stream_data *stream;
+
+	stream = kzalloc(sizeof(*stream), GFP_KERNEL);
+	if (!stream)
+		return -ENOMEM;
+
+	stream->sdw_stream = (struct sdw_stream_runtime *)sdw_stream;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: no need for a type-cast

File: sound/soc/codecs/rt715-sdca.c [hunk 31]
```diff
@@ -0,0 +1,943 @@
...
+		return -EINVAL;
+	}
+
+	stream_config.frame_rate =  params_rate(params);
+	stream_config.ch_count = params_channels(params);
+	stream_config.bps = snd_pcm_format_width(params_format(params));
+	stream_config.direction = direction;
+
+	num_channels = params_channels(params);
+	port_config.ch_mask = (1 << (num_channels)) - 1;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: superfluous parentheses aroung `num_channels`

File: sound/soc/codecs/rt715-sdca.c [hunk 32]
```diff
@@ -0,0 +1,943 @@
...
+	port_config.num = port;
+
+	retval = sdw_stream_add_slave(rt715->slave, &stream_config,
+					&port_config, 1, stream->sdw_stream);
+	if (retval) {
+		dev_err(dai->dev, "Unable to configure port\n");
+		return retval;
+	}
+
+
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: 1 empty line

File: sound/soc/codecs/rt715-sdca.c [hunk 33]
```diff
@@ -0,0 +1,943 @@
...
+		val |= 0xe;
+		break;
+	case 768000:
+		val |= 0xf;
+		break;
+	default:
+		dev_err(component->dev, "Unsupported sample rate %d\n",
+			params_rate(params));
+		return -EINVAL;
+	}
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: you can remove `val` initialisation and just use `val = 0x..` above

File: sound/soc/codecs/rt715-sdca.c [hunk 34]
```diff
@@ -0,0 +1,943 @@
...
+#include <linux/of_gpio.h>
+#include <linux/of_device.h>
+#include <sound/core.h>
+#include <sound/pcm.h>
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+#include <sound/soc-dapm.h>
+#include <sound/initval.h>
+#include <sound/tlv.h>
+#include <sound/hda_verbs.h>
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: I don't see any use of gpio, i2c, regulator or of_ headers? also delay.h, hda_verbs.h? Please, verify, that all headers are really needed.

File: sound/soc/codecs/rt715-sdca.c [hunk 35]
```diff
@@ -0,0 +1,943 @@
...
+
+	pm_runtime_mark_last_busy(&slave->dev);
+	pm_runtime_put_autosuspend(&slave->dev);
+
+	return 0;
+
+}
+
+MODULE_DESCRIPTION("ASoC rt715 driver");
+MODULE_DESCRIPTION("ASoC rt715 driver SDW");
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.c]: Don't think you need two `MODULE_DESCRIPTION()` lines

File: sound/soc/codecs/rt715-sdca.h [hunk 36]
```diff
@@ -0,0 +1,124 @@
...
+// 
+// rt715-sdca.h -- RT715 ALSA SoC audio driver header
+// 
+// Copyright(c) 2019 Realtek Semiconductor Corp.
+// 
+
+#ifndef __RT715_SDCA_H__
+#define __RT715_SDCA_H__
+
+#include <linux/regulator/consumer.h>
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.h]: you include only one header and it's the one you don't need :-) Please, include all needed headers (e.g. workqueue.h, sdw.h) and forward-declare all structs, that aren't defined in those headers or here.

File: sound/soc/codecs/rt715-sdca.h [hunk 37]
```diff
@@ -0,0 +1,124 @@
...
+	enum sdw_slave_status status;
+	struct sdw_bus_params params;
+	bool hw_init;
+	bool first_init;
+	int l_is_unmute;
+	int r_is_unmute;
+	int hw_sdw_ver;
+};
+
+struct sdw_stream_data {
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdca.h]: don't think it's a good idea to define a struct with a generic name in a codec-specific header.

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 38]
```diff
@@ -0,0 +1,479 @@
...
+#include <linux/device.h>
+#include <linux/mod_devicetable.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/regmap.h>
+#include <sound/soc.h>
+#include "rt715-sdca.h"
+#include "rt715-sdw-sdca.h"
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdw-sdca.c]: please, sanitise: e.g. of.h doesn't seem to be needed, possibly also delay.h

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 39]
```diff
@@ -0,0 +1,479 @@
...
+		if (ret < 0)
+			return ret;
+		ret = regmap_read(rt715->sdw_regmap,
+			reg | BIT(13), &sdw_data_1);
+		if (ret < 0)
+			return ret;
+		*val = ((sdw_data_1 & 0xff) << 8) | (sdw_data_0 & 0xff);
+		goto _done_;
+	default:
+		break;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdw-sdca.c]: you can just put the code below up to _done_ here under `default` and remove the `goto`

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 40]
```diff
@@ -0,0 +1,479 @@
...
+		CH_07):
+	case RT715_SDCA_ADDR(FUNC_02, FU_DMIC_GAIN_EN, FU_DMIC_GAIN_CTRL,
+		CH_08):
+		regmap_write(rt715->sdw_regmap,
+			reg | BIT(13), ((val >> 8) & 0xff));
+		regmap_write(rt715->sdw_regmap,
+			reg, (val & 0xff));
+		goto _done_;
+	default:
+		break;
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdw-sdca.c]: same here

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 41]
```diff
@@ -0,0 +1,479 @@
...
+	case RT715_SDCA_ADDR(FUNC_02, FU_DMIC_GAIN_EN, FU_DMIC_GAIN_CTRL,
+		CH_05):
+	case RT715_SDCA_ADDR(FUNC_02, FU_DMIC_GAIN_EN, FU_DMIC_GAIN_CTRL,
+		CH_06):
+	case RT715_SDCA_ADDR(FUNC_02, FU_DMIC_GAIN_EN, FU_DMIC_GAIN_CTRL,
+		CH_07):
+	case RT715_SDCA_ADDR(FUNC_02, FU_DMIC_GAIN_EN, FU_DMIC_GAIN_CTRL,
+		CH_08):
+		regmap_write(rt715->sdw_regmap,
+			reg | BIT(13), ((val >> 8) & 0xff));
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdw-sdca.c]: superfluous parentheses around `(val >> 8) & 0xff`

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 42]
```diff
@@ -0,0 +1,479 @@
...
+		regmap_write(rt715->sdw_regmap,
+			reg, (val & 0xff));
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdw-sdca.c]: superfluous parentheses around `val & 0xff`

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 43]
```diff
@@ -0,0 +1,479 @@
...
+		goto _done_;
+	default:
+		break;
+	}
+
+	/* SDCA mapping case or normal address */
+	if (reg & 0x40000000 || reg < 0xffff) {
+		regmap_write(rt715->sdw_regmap,
+			reg, val);
+	}
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdw-sdca.c]: superfluous braces. You can also fit the `regmap_write()` onto one line, there were a couple more similar unnecessary line-breaks above

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 44]
```diff
@@ -0,0 +1,479 @@
...
+	if (!sdw_regmap)
+		return -EINVAL;
+
+	regmap = devm_regmap_init(&slave->dev, NULL,
+		&slave->dev, &rt715_sdca_regmap);
+
+	if (!regmap)
+		return -EINVAL;
+
+	rt715_init(&slave->dev, sdw_regmap, regmap, slave);
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdw-sdca.c]: you wanted `return rt715_init()`?

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 45]
```diff
@@ -0,0 +1,479 @@
...
+	if (!time) {
+		dev_err(&slave->dev, "Enumeration not complete, timed out\n");
+		return -ETIMEDOUT;
+	}
+	regcache_cache_only(rt715->regmap, false);
+
+	ret = regcache_sync_region(rt715->regmap, 0x2000000, 0x61020ff);
+	ret = regcache_sync_region(rt715->regmap, 0x40400000, 0x40880f80);
+
+
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdw-sdca.c]: 1 empty line

File: sound/soc/codecs/rt715-sdw-sdca.c [hunk 46]
```diff
@@ -0,0 +1,479 @@
...
+	SET_SYSTEM_SLEEP_PM_OPS(rt715_dev_suspend, rt715_dev_resume)
+	SET_RUNTIME_PM_OPS(rt715_dev_suspend, rt715_dev_resume, NULL)
+};
+
+static struct sdw_driver rt715_sdw_driver = {
+	.driver = {
+		   .name = "rt715",
+		   .owner = THIS_MODULE,
+		   .pm = &rt715_pm,
+		   },
```
Comment (lyakh) [inline sound/soc/codecs/rt715-sdw-sdca.c]: one TAB too many above

**Pull request closed without merging**